File: src/main/java/render/FontRenderer.java
package render;

import gl.*;
import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.lwjgl.opengl.GL31;
import org.lwjgl.system.MemoryStack;
import util.FloatList;

import java.io.File;
import java.io.IOException;

public class FontRenderer {

    private final int vao;
    private final int vbo;
    public Font font;
    private int quads;
    private FloatList verts;

    public float fontHeight;
    public float[] fontWidths;

    public ShaderProgram fontShader;
    public int fontProj;

    public static final Vector3f SHADOW = new Vector3f(0.45f, 0.45f, 0.45f);
    public static final Vector3f WHITE = new Vector3f(1.0f, 1.0f, 1.0f);
    public static final Vector3f YELLOW = new Vector3f(1.0f, 1.0f, 0.0f);
    public static final Vector3f INACTIVE = new Vector3f(0.75f, 0.75f, 0.75f);
    public char[] char_buffer;

    public static FontRenderer FONT_RENDERER = null;

    static {
        try {
            FONT_RENDERER = new FontRenderer();
        } catch (IOException e) {
            System.err.println(e.getMessage());
            System.exit(-1);
        }
    }

    public FontRenderer() throws IOException {
        vao = GL31.glGenVertexArrays();
        GL31.glBindVertexArray(vao);
        vbo = GL31.glGenBuffers();
        GL31.glBindBuffer( GL31.GL_ARRAY_BUFFER, vbo);
        font = new Font(new File("font", "alphabet.png"));

        fontHeight = (float)font.getHeight();
        fontWidths = new float[Font.FONT_TABLE.length()];

        for(int i = 0;i < Font.FONT_TABLE.length();i++) {
            fontWidths[i] = font.getWidthOf(Font.FONT_TABLE.charAt(i));
        }

        VertexShader textVert= new VertexShader(new File("shader", "text.vert"));
        FragmentShader textFrag = new FragmentShader(new File("shader", "text.frag"));
        fontShader = new ShaderProgram(textVert, textFrag);

        fontShader.use();

        fontProj = GL31.glGetUniformLocation(fontShader.getProgram(), "proj");

        int posAttrib = GL31.glGetAttribLocation(fontShader.getProgram(), "position");
        GL31.glEnableVertexAttribArray(posAttrib);
        GL31.glVertexAttribPointer(posAttrib, 2, GL31.GL_FLOAT, false, 7 * 4, 0L);

        int colorAttrib = GL31.glGetAttribLocation(fontShader.getProgram(), "color");
        GL31.glEnableVertexAttribArray(colorAttrib);
        GL31.glVertexAttribPointer(colorAttrib, 3, GL31.GL_FLOAT, false, 7 * 4, (2 * 4));

        int texAttrib = GL31.glGetAttribLocation(fontShader.getProgram(), "texcoord");
        GL31.glEnableVertexAttribArray(texAttrib);
        GL31.glVertexAttribPointer(texAttrib, 2, GL31.GL_FLOAT, false, 7 * 4, (5 * 4));

        verts = new FloatList();
    }

    public void renderWithShadow(float x, float y, String text, float scale, Vector3f color) {
        renderText(x+(1.0f),y+(1.0f),text,scale,SHADOW);
        renderText(x,y,text,scale,color);
    }

    public void renderWithShadow(float x, float y, String text, float scale) {
        renderWithShadow(x,y,text,scale,WHITE);
    }

    public void renderText(float x, float y, String text, float scale, Vector3f color) {
        GL31.glBindBuffer(GL31.GL_ARRAY_BUFFER, vbo);

        float cX = x;
        float cY = y;
        float height = fontHeight * scale;
        char_buffer = text.toCharArray();

        for (char c : char_buffer) {
            if (c == '\n') {
                cY += height + scale;
                cX = x;
                continue;
            }

            float width = getFontWidth(c) * scale;

            verts.append(cX);
            verts.append(cY);
            verts.append(color.x);
            verts.append(color.y);
            verts.append(color.z);
            verts.append(font.getUVOf(c, UVPosition.U1));
            verts.append(font.getUVOf(c, UVPosition.V1));

            verts.append(cX);
            verts.append(cY + height);
            verts.append(color.x);
            verts.append(color.y);
            verts.append(color.z);
            verts.append(font.getUVOf(c, UVPosition.U1));
            verts.append(font.getUVOf(c, UVPosition.V2));

            verts.append(cX + width);
            verts.append(cY + height);
            verts.append(color.x);
            verts.append(color.y);
            verts.append(color.z);
            verts.append(font.getUVOf(c, UVPosition.U2));
            verts.append(font.getUVOf(c, UVPosition.V2));

            verts.append(cX + width);
            verts.append(cY + height);
            verts.append(color.x);
            verts.append(color.y);
            verts.append(color.z);
            verts.append(font.getUVOf(c, UVPosition.U2));
            verts.append(font.getUVOf(c, UVPosition.V2));

            verts.append(cX + width);
            verts.append(cY);
            verts.append(color.x);
            verts.append(color.y);
            verts.append(color.z);
            verts.append(font.getUVOf(c, UVPosition.U2));
            verts.append(font.getUVOf(c, UVPosition.V1));

            verts.append(cX);
            verts.append(cY);
            verts.append(color.x);
            verts.append(color.y);
            verts.append(color.z);
            verts.append(font.getUVOf(c, UVPosition.U1));
            verts.append(font.getUVOf(c, UVPosition.V1));

            cX += width + scale;
        }

        quads = verts.getLength() / 7;
    }

    public void renderWithShadowImmediate(Matrix4f proj, float x, float y, String text, float scale, Vector3f color) {
        renderWithShadow(x,y,text,scale,color);
        draw(proj);
    }

    public void renderWithShadowImmediate(Matrix4f proj, float x, float y, String text, float scale) {
        renderWithShadowImmediate(proj, x, y, text, scale, WHITE);
    }

    public float getFontWidth(char c) {
        if(Font.FONT_TABLE.indexOf(c) != -1) {
            return fontWidths[Font.FONT_TABLE.indexOf(c)];
        } else {
            return font.getWidthOf(c);
        }
    }

    public float getStringWidth(String s, float scale) {
        float width = 0.0f;
        char_buffer = s.toCharArray();
        for(char c : char_buffer) {
            width += (getFontWidth(c) * scale);
            width += scale;
        }
        return width;
    }

    public void draw(Matrix4f proj) {
        if(verts.getLength() == 0)
            return;
        fontShader.use();
        font.use();
        GL31.glBindVertexArray(vao);
        GL31.glBindBuffer( GL31.GL_ARRAY_BUFFER, vbo);
        GL31.glBufferData( GL31.GL_ARRAY_BUFFER, verts.getStore(),  GL31.GL_DYNAMIC_DRAW);

        MemoryStack stack = null;
        try {
            stack = MemoryStack.stackPush();
            GL31.glUniformMatrix4fv(fontProj, false, proj.get(stack.mallocFloat(16)));
        } finally {
            stack.pop();
        }
        GL31.glDrawArrays(GL31.GL_TRIANGLES, 0, quads);
        verts.clear();
    }
}

File: src/main/kotlin/render/FontRenderer.kt
package render

import gl.*
import org.joml.Matrix4f
import org.joml.Vector3f
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL31.*
import org.lwjgl.system.MemoryStack
import java.io.File


object FontRenderer {
    var vao: Int
    var vbo: Int
    var font: Font
    var quads: Int = 0
    val verts = mutableListOf<Float>()
    var baked_verts = floatArrayOf()

    val fontHeight: Float
    val fontWidths: FloatArray

    var fontShader: ShaderProgram
    var fontProj: Int

    val SHADOW = Vector3f(0.45f, 0.45f, 0.45f)
    val WHITE = Vector3f(1.0f, 1.0f, 1.0f)

    init {
        vao = glGenVertexArrays()
        glBindVertexArray(vao)
        vbo = glGenBuffers()
        glBindBuffer(GL_ARRAY_BUFFER, vbo)
        font = Font(File("font", "alphabet.png"))

        fontHeight = font.height.toFloat()
        fontWidths = FloatArray(font.fontTable.length) {
            font.getWidthOf(font.fontTable[it]).toFloat()
        }

        var textVert= VertexShader(File("shader", "text.vert"))
        var textFrag = FragmentShader(File("shader", "text.frag"))
        fontShader = ShaderProgram(textVert, textFrag)

        fontShader.use()

        fontProj = glGetUniformLocation(fontShader.program, "proj")

        var posAttrib = glGetAttribLocation(fontShader.program, "position")
        glEnableVertexAttribArray(posAttrib)
        glVertexAttribPointer(posAttrib, 2, GL_FLOAT, false, 7 * 4, 0L)

        var colorAttrib = glGetAttribLocation(fontShader.program, "color")
        glEnableVertexAttribArray(colorAttrib)
        glVertexAttribPointer(colorAttrib, 3, GL_FLOAT, false, 7 * 4, (2 * 4))

        var texAttrib = glGetAttribLocation(fontShader.program, "texcoord")
        glEnableVertexAttribArray(texAttrib)
        glVertexAttribPointer(texAttrib, 2, GL_FLOAT, false, 7 * 4, (5 * 4))
    }

    fun renderWithShadow(x: Float, y: Float, text: String, scale: Float, color: Vector3f = WHITE) {
        renderText(x+(1.0f),y+(1.0f),text,scale,SHADOW)
        renderText(x,y,text,scale,color)
    }

    fun renderText(x: Float, y: Float, text: String, scale: Float, color: Vector3f = WHITE) {
        glBindBuffer(GL_ARRAY_BUFFER, vbo)

        var cX = x
        var cY = y
        val height = fontHeight * scale

        for (c in text) {
            if (c == '\n') {
                cY += height + scale
                cX = x
                continue
            }

            val width = getFontWidth(c) * scale

            verts.add(cX)
            verts.add(cY)
            verts.add(color.x)
            verts.add(color.y)
            verts.add(color.z)
            verts.add(font.getUVOf(c, UVPosition.U1))
            verts.add(font.getUVOf(c, UVPosition.V1))

            verts.add(cX)
            verts.add(cY + height)
            verts.add(color.x)
            verts.add(color.y)
            verts.add(color.z)
            verts.add(font.getUVOf(c, UVPosition.U1))
            verts.add(font.getUVOf(c, UVPosition.V2))

            verts.add(cX + width)
            verts.add(cY + height)
            verts.add(color.x)
            verts.add(color.y)
            verts.add(color.z)
            verts.add(font.getUVOf(c, UVPosition.U2))
            verts.add(font.getUVOf(c, UVPosition.V2))

            verts.add(cX + width)
            verts.add(cY + height)
            verts.add(color.x)
            verts.add(color.y)
            verts.add(color.z)
            verts.add(font.getUVOf(c, UVPosition.U2))
            verts.add(font.getUVOf(c, UVPosition.V2))

            verts.add(cX + width)
            verts.add(cY)
            verts.add(color.x)
            verts.add(color.y)
            verts.add(color.z)
            verts.add(font.getUVOf(c, UVPosition.U2))
            verts.add(font.getUVOf(c, UVPosition.V1))

            verts.add(cX)
            verts.add(cY)
            verts.add(color.x)
            verts.add(color.y)
            verts.add(color.z)
            verts.add(font.getUVOf(c, UVPosition.U1))
            verts.add(font.getUVOf(c, UVPosition.V1))

            cX += width + scale
        }

        quads = verts.size / 7
    }

    fun renderWithShadowImmediate(proj: Matrix4f, x: Float, y: Float, text: String, scale: Float, color: Vector3f = WHITE) {
        renderWithShadow(x,y,text,scale,color)
        draw(proj)
    }

    fun getFontWidth(c: Char): Float {
        if(c in font.fontTable) {
            return fontWidths[font.fontTable.indexOf(c)]
        } else {
            return font.getWidthOf(c).toFloat()
        }
    }


    fun getStringWidth(s: String, scale: Float): Float {
        var width = 0.0f
        for(c in s) {
            width += (getFontWidth(c) * scale)
            width += scale
        }
        return width
    }

    fun bake_font() {
        if(verts.size > baked_verts.size){
            baked_verts = FloatArray(verts.size)
            System.out.println("resizing font")
        }
        for(i in 0 until verts.size) {
            baked_verts[i] = verts[i]
        }
    }

    fun clear_all() {
        for(i in baked_verts.indices) {
            baked_verts[i] = 0.0f
        }
        verts.clear()
    }

    fun draw(proj: Matrix4f) {
        if(verts.size == 0)
            return
        fontShader.use()
        font.use()
        glBindVertexArray(vao)
        glBindBuffer(GL_ARRAY_BUFFER, vbo)
        bake_font()
        glBufferData(GL_ARRAY_BUFFER, baked_verts, GL_DYNAMIC_DRAW)


        var stack: MemoryStack? = null
        try {
            stack = MemoryStack.stackPush()
            glUniformMatrix4fv(fontProj, false, proj.get(stack.mallocFloat(16)))
        } finally {
            stack?.pop()
        }
        glDrawArrays(GL_TRIANGLES, 0, quads)
        clear_all()
    }
}